### ISSUE TITLE
Reading controls in responses from the LDAP server

### ISSUE THREAD

**Issue by AntonNeld** [ldapts/ldapts#106]
Hi,

(I'm not very experienced with LDAP, so I may get some terminology wrong here. I hope it's clear enough anyway.)

I am trying to use an extension that ldapts has no knowledge of ([this](https://ldapwiki.com/wiki/Draft-behera-ldap-password-policy), but the details shouldn't matter too much). The client is supposed to send a control with the bind request, and receive a control back from the server with some extra info. I can send the control without issue, but I can't figure out how to get at the one in the response.

I've tried to subclass Control, but parseControl() does not seem to be executed:

```js
import fsp from 'fs/promises';
import ldapts from 'ldapts';
import { Control } from 'ldapts/controls/index.js';

class MyControl extends Control {
  constructor() {
    // This OID is the same for both the request and the response in my case,
    // but I don't know if that holds in general.
    super('1.3.6.1.4.1.42.2.27.8.5.1');
  }

  parseControl(reader) {
    console.log('reading');
  }

  writeControl(writer) {
    console.log('writing');
  }
}

async function main() {
  const client = new ldapts.Client({
    url: 'ldaps://localhost:8636',
    tlsOptions: {
      rejectUnauthorized: false,
    },
  });

  const control = new MyControl();
  await client.bind(
    'uid=someuser,ou=people,dc=la,dc=adp,dc=ericsson',
    'somepassword',
    control
  );
  await client.unbind();
}

main(); // Prints only 'writing'
```

From some console.log debugging, it seems like the control in the response is discarded [here](https://github.com/ldapts/ldapts/blob/d9146a5aa6d60a399bb895599e1764a37727264a/src/ControlParser.ts#L32) due to it not being one of the types ldapts knows about.

I have tried this with ldapts 3.0.5, and an OpenLDAP server (not sure which version, but I can find out if it's relevant) with the 'ppolicy' extension enabled.

These are my questions:
- Is there a way to read an arbitrary control sent from the server?
- Assuming that way is not through parseControl(), what is parseControl() used for?

**Comment by jgeurts** [comment]
Sorry for the long delay on getting back to you about this. It looks to be a bug in the design of ldapts. Currently, messages are received from the socket, parsed, and then the parsed message is emitted back to the Client class for final handling. I'm thinking that the message parsing will need to be deferred until after the message is passed back to the Client class. That way, we'll know what type of message is being processed and can try parsing any additional controls specified. I'm open to other ideas, too!

**Comment by AntonNeld** [comment]
No problem :)

That sounds like a good solution to me.

Would the "additional controls specified" be the array of controls passed to (e.g.) bind(), like today? That would certainly work for my use-case, but are there are controls that the client shouldn't send in the request but that can be present in the response? If so, I guess there should be a way to specify which Control instances should be sent and which are only used for parsing.

**Comment by thernstig** [comment]
@jgeurts Any chance you got time for this one 😆 ?

**Comment by thernstig** [comment]
@jgeurts would this be hard to implement?

**Comment by jgeurts** [comment]
I have a plan for how to support this. Will try to work on it in the next week or two.

**Comment by jgeurts** [comment]
Could you please try `4.3.0-beta1`? 

@TimoHocker If you have a chance, could you please verify that I didn't break SASL bind responses with the `4.3.0-beta1` release?

**Comment by TimoHocker** [comment]
@jgeurts My tests are still passing. `4.3.0-beta1` looks good to me.

**Comment by jgeurts** [comment]
@TimoHocker Thank you!

**Comment by thernstig** [comment]
@jgeurts intending to test this soon! Just have had some very urgent issues the past week. Going to look on it next week!

**Comment by thernstig** [comment]
@jgeurts I amended my existing code with this (left out parts I did not change)

```js
import ldapts, { Control, NoSuchObjectError } from 'ldapts'; // added import of 'Control'

class MyControl extends Control {
  constructor() {
    // This OID is the same for both the request and the response in my case,
    // but I don't know if that holds in general.
    super('1.3.6.1.4.1.42.2.27.8.5.1');
  }

  parseControl(reader) {
    console.log('reading');
  }

  writeControl(writer) {
    console.log('writing');
  }
}
const control = new MyControl();

await client.bind('EXTERNAL', undefined, control); // here I only added the 'control' part
```

This gave:
```sh
SaslBindInProgressError: SASL(0): successful result:  Code: 0xe
       at StatusCodeParser.parse (/home/user1/code/projectA/node_modules/ldapts/StatusCodeParser.js:28:24)
       at Client._sendBind (/home/user1/code/projectA/node_modules/ldapts/Client.js:472:55)
       at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
       at async Client.bindSASL (/home/user1/code/projectA/node_modules/ldapts/Client.js:168:9)
       at async Client.bind (/home/user1/code/projectA/node_modules/ldapts/Client.js:141:13)
       at async getUserDetails (file:///home/user1/code/projectA/framework/authentication/ldapAdp.js:210:7)
       at async Strategy._verify (file:///home/user1/code/projectA/framework/authentication/ldapAdp.js:100:13) {
     code: 14,
     response: BindResponse {
       version: 3,
       messageId: 2,
       controls: [],
       status: 14,
       matchedDN: '',
       errorMessage: 'SASL(0): successful result: ',
       data: [Array],
       protocolOperation: 97
     }
   }
 }
```

Is the problem here that when using SASL I have to specify `undefined` as the second parameter to `client.bind()`?
If so, how would I get around that?

**Comment by jgeurts** [comment]
Thanks for letting me know, I'll get that sorted out!

**Comment by thernstig** [comment]
@jgeurts I am aware I am asking for a lot, but if you do get a chance to look at it again it would be highly appreciated. There is so much power in being able to read controls e.g. OpenLDAP Password Policy (ppolicy). And of course many more.

References:
https://linux.die.net/man/5/slapo-ppolicy
https://tobru.ch/openldap-password-policy-overlay/
https://www.zytrax.com/books/ldap/ch6/ppolicy.html

**Comment by jgeurts** [comment]
Hi @thernstig, I apologize for the delay on this. I'm trying to get an openldap instance docker sidecar running since it's pretty hard to debug otherwise. I was able to get https://github.com/ldapjs/docker-test-openldap hooked up pretty easily, but configuring that for SASL is still a step I could use help with. 

It looks like it's a sasl connection issue and maybe not related to the undefined password. I'm not sure if the additional controls are contributing to the issue or not. Curious, does calling bind again help with the issue?

**Comment by thernstig** [comment]
Hi @jgeurts. I will try some more next week. I want to verify I did not do it wrong using SASL when using the control.

The problem for me is that I get an OpenLDAP server handed to me already prepared with SASL. So I have limited to know experience on that. Even my LDAP knowledge is junior at best. I can check though if I can find something, maybe I can find out some info on how to config SASL.

**Comment by thernstig** [comment]
@jgeurts I have found some "good news". I tested again and noticed that the entire SASL EXTERNAL flow broke with ldapts version 4.2.0. 4.1.1 works fine, so something in the 4.2.0 update broken the normal flow. (I noticed I was pinged in that PR but must have missed it). I will create a new issue about this.

**Comment by thernstig** [comment]
@jgeurts I have worked around the error reported in #133, so that is not an issue anymore.

I tried the control parsing in `4.3.0-beta1.` I am using a control for a `.search()`:

```js
class MyControl extends ldapts.Control {
  constructor() {
    super('1.3.6.1.4.1.42.2.27.8.5.1');
  }

  parseControl(reader) {
    console.log('reading');
  }

  writeControl(writer) {
    console.log('writing');
  }
}
const control = new MyControl();
```

I can see the `writing` but not the `reading`. I am supposed to do that right, otherwise I have no idea how to parse the response?

**Comment by thernstig** [comment]
@jgeurts if you want to test this in a test case later after #135 is merged, here is the description of the ppolicy control:

https://datatracker.ietf.org/doc/html/draft-behera-ldap-password-policy-11#name-response-control

**Comment by jgeurts** [comment]
Thanks @thernstig. I was able to confirm the `parseControl` is not being called. I'm trying to work out a fix for that.

**Comment by jgeurts** [comment]
This should be fixed with v5.0.0
